source/extensions/filters/http/common/aws/signer_impl.h

```diff
@@ -0,0 +1,94 @@
+#pragma once
+
+#include "common/common/logger.h"
+#include "common/common/utility.h"
+#include "common/singleton/const_singleton.h"
+
+#include "extensions/filters/http/common/aws/credentials_provider.h"
+#include "extensions/filters/http/common/aws/signer.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Common {
+namespace Aws {
+
+class SignatureHeaderValues {
+public:
+  const Http::LowerCaseString ContentSha256{"x-amz-content-sha256"};
+  const Http::LowerCaseString Date{"x-amz-date"};
+  const Http::LowerCaseString SecurityToken{"x-amz-security-token"};
+};
+
+typedef ConstSingleton<SignatureHeaderValues> SignatureHeaders;
+
+class SignatureConstantValues {
+public:
+  const std::string Aws4Request{"aws4_request"};
+  const std::string AuthorizationHeaderFormat{
+      "AWS4-HMAC-SHA256 Credential={}/{}, SignedHeaders={}, Signature={}"};
+  const std::string CredentialScopeFormat{"{}/{}/{}/aws4_request"};
+  const std::string HashedEmptyString{
+      "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"};
+  const std::string SignatureVersion{"AWS4"};
+  const std::string StringToSignFormat{"AWS4-HMAC-SHA256\n{}\n{}\n{}"};
+
+  const std::string LongDateFormat{"%Y%m%dT%H%M00Z"};
+  const std::string ShortDateFormat{"%Y%m%d"};
+};
+
+typedef ConstSingleton<SignatureConstantValues> SignatureConstants;
+
+/**
+ * Implementation of the Signature V4 signing process.
+ * See https://docs.aws.amazon.com/general/latest/gr/signature-version-4.html
+ */
+class SignerImpl : public Signer, public Logger::Loggable<Logger::Id::aws> {
+public:
+  SignerImpl(absl::string_view service_name, absl::string_view region,
+             const CredentialsProviderSharedPtr& credentials_provider, TimeSource& time_source)
+      : service_name_(service_name), region_(region), credentials_provider_(credentials_provider),
+        time_source_(time_source), long_date_formatter_(SignatureConstants::get().LongDateFormat),
+        short_date_formatter_(SignatureConstants::get().ShortDateFormat) {}
+
+  void sign(Http::Message& message) override;
+
+private:
+  friend class SignerImplTest;
+
+  const std::string service_name_;
+  const std::string region_;
+  CredentialsProviderSharedPtr credentials_provider_;
+  TimeSource& time_source_;
+  DateFormatter long_date_formatter_;
+  DateFormatter short_date_formatter_;
+
+  std::string createContentHash(Http::Message& message) const;
```

Comment:
Nit: generally we prefer methods before data members in each access scoped section, e.g. private here.